ddtrace/internal/module.py

```diff
@@ -0,0 +1,454 @@
+from collections import defaultdict
+from os.path import abspath
+from os.path import expanduser
+from os.path import isdir
+from os.path import isfile
+from os.path import join
+import sys
+from types import ModuleType
+from typing import Any
+from typing import Callable
+from typing import DefaultDict
+from typing import Dict
+from typing import Iterator
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import Tuple
+from typing import Union
+
+from ddtrace.internal.compat import PY2
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils import get_argument_value
+
+
+log = get_logger(__name__)
+
+HookType = Callable[[ModuleType, Any], None]
+
+
+_run_code = None
+_post_run_module_hooks = []
+
+
+def _wrapped_run_code(*args, **kwargs):
+    # type: (*Any, **Any) -> Dict[str, Any]
+    global _run_code, _post_run_module_hooks
+
+    # DEV: If we are calling this wrapper then _run_code must have been set to
+    # the original runpy._run_code.
+    assert _run_code is not None
+
+    mod_name = get_argument_value(args, kwargs, 3, "mod_name")
+
+    try:
+        return _run_code(*args, **kwargs)
+    finally:
+        module = sys.modules[mod_name]
+        for hook in _post_run_module_hooks:
+            hook(module)
+
+
+def _patch_run_code():
+    # type: () -> None
+    global _run_code
+
+    if _run_code is None:
+        import runpy
+
+        _run_code = runpy._run_code  # type: ignore[attr-defined]
+        runpy._run_code = _wrapped_run_code  # type: ignore[attr-defined]
+
+
+def register_post_run_module_hook(hook):
```

Comment:
We don't have tests for this functionality and it is independent of the rest of the module, should we do it as a follow-up @P403n1x87?

Reply:
This is tested by the new `test_post_run_module_hook` test case.